File: dev-c++/0007-reverse-integer.cpp
class Solution {
public:
    int reverse(int _x) {
        bool sign = (_x > 0);        
        unsigned int x = _x;         
        unsigned int result = 0;

        if(! sign) { x = -x; }

        while(x != 0) {
            if(result > INT_MAX / 10) { return 0; }

            result *= 10;
            result += (x % 10);
            x /= 10;
        }

        return (sign) ? (int) result : -((int) result);
    }
};

File: dev-rust/0007-reverse-integer.rs
impl Solution {
    pub fn reverse(x: i32) -> i32 {
        if x == i32::MIN { return 0; }
        if x == 0        { return 0; }

        let is_negative = (x < 0);
        let mut remainder: i32 = x.abs();
        let mut result: i32 = 0;

        while(remainder != 0) {
            result = match result.checked_mul(10) { 
                Some(n) => n, 
                None    => return 0 
            };

            result = match result.checked_add(remainder % 10) { 
                Some(n) => n, 
                None    => return 0 
            };

            remainder /= 10;
        }

        if is_negative { return -result; }
        else           { return  result; }
    }
}
